docs/shared/workspace/project-graph-plugins.md

```diff
@@ -66,6 +68,14 @@ export function processProjectGraph(
 }
 ```
 
+## Adding New Dependencies to the Project Graph
+
+Project Graph Plugins can be used to add smarter dependency resolution to projects already in the workspace. Projects in the workspace are first party code whose dependencies change as the code in the workspace changes and matter to Nx the most. Such projects should be defined in `workspace.json` and `nx.json` and will be automatically included as nodes in the project graph. However, when some projects are written in other languages, the relationships between these projects will not be clear to Nx out of the box. A Project Graph Plugin can be used to add these relationships. Note that even though the plugin is written in JavaScript, resolving dependencies of different languages will probably be more easily written in their native language. Therefore, a common approach is to spawn a new process and communicate via IPC or `stdout`.
+
+## Adding New Nodes to the Project Graph
+
+Sometimes it can be valuable to have third party packages as part of the project graph. A Project Graph Plugin can be used to add these packages to the project graph. After these packages are added as nodes to the project graph, dependencies can then be drawn from the workspace projects to the third party packages.
```

Comment:
> A Project Graph Plugin can be used to add these packages to the project graph.

This sentence is repeated. Shouldn't we have that somehow in the intro of this doc article? Like explaining why it is useful to have a project graph plugin etc...

Also, shouldn't these sections `Adding New Dependencies...` and `Adding New Nodes...` be before the actual `Implementing a Project Graph Processor` which shows the concrete way of adding nodes and dependencies in code?  
alternatively we could just have a small excerpt in the code example showing how to implement the processor and then further snippets in these sections that add the pieces of `builder.addDependency` and `builder.addNodes`

Reply:
I rewrote the docs a bit. Let me know if the new version makes more sense. :)

Reply:
LGTM now :)